### PR TITLE
Extractor unit model for minerals processing

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This is a draft of the first unit model for PrOMMiS. It is a generic model for a multi-stream contactor unit which we hope will form the basis for a large number of more specific unit models in the future.

Docs will come once I have some feedback on the model.

## Changes proposed in this PR:
- Add new Extractor unit model and tests
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by adowling2 — COMMENTED** [review]
Looks great. We should talk about extending this to more than 2 components for separation.

**Comment by adowling2** [comment]
@mollydougher Here is a draft Pyomo implementation of the diafiltration model (3 stages for the test)

File: idaes/models/unit_models/tests/test_extractor.py [hunk 1]
```diff
@@ -0,0 +1,2711 @@
...
+    def test_diafiltration_build(self, model):
+        # TODO: More checks here
+        assert isinstance(model.fs.stage3.retentate_inlet, Port)
+        assert isinstance(model.fs.stage3.retentate_outlet, Port)
+        assert not hasattr(model.fs.stage3, "permeate_inlet")
+        assert isinstance(model.fs.stage3.permeate_outlet, Port)
+
+    @pytest.mark.integration
+    def test_initialize_and_solve(self, model):
+        # Start with stage 3
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/unit_models/tests/test_extractor.py:2519]: Will do.

**Comment by andrewlee94** [comment]
@adowling2 So, in theory at least, this base model will handle as many components and streams as we want (so it could in theory do a multi-component combined solvent extraction and leaching process). We just need a test case to try out (and the necessary separation/performance equations).

File: docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+* states for feed blocks for each stream,
+* material transfer terms (finite elements :math:`\times` interacting streams :math:`\times` common components),
+* energy transfer terms if included (finite elements :math:`\times` interacting streams)
+* pressure change terms if included (finite elements :math:`\times` streams with pressure change)
+
+Model Structure
+---------------
+
+Due to the custom nature of multi-stream contactors, this model does not make use of control volumes. Instead, a set of StateBlocks (named using the name given in the ``streams`` configuration dictionary) are created for each stream indexed by time and the set of finite elements, with an additional StateBlock for the feed state (named using the stream name appended with ``_inlet_state``) indexed only by time (unless ``has_feed`` is set to ``False`` for that stream). For streams with side streams (feed or draw), an additional set of indexed StateBlocks (named using the stream name appended with ``_side_stream_state``) is created for the side states which are indexed by time and the set of side states for that stream.
```
Comment (agarciadiego) [inline docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst:49]: dictionary) **is** created

File: docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+Constraints
+-----------
+
+In all cases, the multi-stage contactor model writes a set of material balances for each stream in the model. For component ``j`` in stream ``s`` the following constraint, named ``stream + "_material_balance"``, is written for all finite elements ``x``:
+
+.. math:: 0 = \sum_p{F_{t,x-,s,p,j}} - \sum_p{F_{t,x,s,p,j}} + \left[ \sum_p{F_{side,t,x,s,p,j}} \right] + \sum_o{M_{t,x,s,o,j}} + \left[ \sum_p{G_{rate,t,x,s,p,j}} + \sum_p{G_{equil,t,x,s,p,j}} + \sum_p{G_{inher,t,x,s,p,j}} \right]
+
+where ``F`` is the material flow term, ``x-`` represents the the previous finite element (``x-1`` in the case of co-current flow and ``x+1`` in the case of counter-current flow), ``F_side`` is the material flow term for a side stream (if present) and ``o`` represents all other streams in the model (for cases where ``s`` is the second index (i.e., M_{t,x,o,s,j}) the term is multiplied by -1). The reaction generation terms are only included if the appropriate reaction type is supported by the reaction or property package for the stream.
```
Comment (agarciadiego) [inline docs/reference_guides/model_libraries/generic/unit_models/mscontactor.rst:82]: the the previous

File: idaes/models/unit_models/mscontactor.py [hunk 4]
```diff
@@ -0,0 +1,1013 @@
...
+    """
+    This is a general purpose sequential-modular Initializer object for
+    multi-stream contactor unit models.
+
+    This routine starts by deactivating any constraints that are not
+    part of the base model and fixing all inter-stream transfer variables.
+    The model is then solved using the Pyomo ssc_solver function
+    to initialize each stream separately.
+
+    The inter-stream transfer variables then unfixed and additional constraints
```
Comment (agarciadiego) [inline idaes/models/unit_models/mscontactor.py:61]: "are" then unfixed

**Review by agarciadiego — APPROVED** [review]
LGTM only a few typos

**Review by eslickj — APPROVED** [review]

**Pull request merged**